### PR TITLE
add test in validation

### PR DESCRIPTION
**Pull request by zjj2wry**
**What this PR does / why we need it**:
#39559

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by shiywang** [comment]
@k8s-bot ok to test

File: pkg/api/validation/validation_test.go [hunk 1]
```diff
@@ -896,6 +896,53 @@ func TestValidateKeyToPath(t *testing.T) {
...
+func TestValidateNFSVolumeSource(t *testing.T) {
+	testCases := []struct {
+		name      string
+		nfs       *api.NFSVolumeSource
+		errtype   field.ErrorType
+		errfield  string
+		errdetail string
+	}{
+		{
+			name:     "unvalid server",
```
Comment (rootfs) [inline pkg/api/validation/validation_test.go]: invalid

File: pkg/api/validation/validation_test.go [hunk 2]
```diff
@@ -896,6 +896,53 @@ func TestValidateKeyToPath(t *testing.T) {
...
+			nfs:      &api.NFSVolumeSource{Server: "", Path: "/tmp"},
+			errtype:  field.ErrorTypeRequired,
+			errfield: "server",
+		},
+		{
+			name:     "unvalid path",
```
Comment (rootfs) [inline pkg/api/validation/validation_test.go]: missing path

**Comment by rootfs** [comment]
@zjj2wry nice work, can you add test cases for glusterfs as well?

**Comment by zjj2wry** [comment]
@rootfs of course, i will do ~

**Comment by zjj2wry** [comment]
@rootfs done

File: pkg/api/validation/validation_test.go [hunk 3]
```diff
@@ -897,6 +897,94 @@ func TestValidateKeyToPath(t *testing.T) {
...
+func TestValidateGlusterfs(t *testing.T) {
+	testCases := []struct {
+		name     string
+		nfs      *api.GlusterfsVolumeSource
+		errtype  field.ErrorType
+		errfield string
+	}{
+		{
+			name:     "missing endpointname",
+			nfs:      &api.GlusterfsVolumeSource{EndpointsName: "", Path: "/tmp"},
```
Comment (rootfs) [inline pkg/api/validation/validation_test.go]: nfs -> glusterfs (and below too)

File: pkg/api/validation/validation_test.go [hunk 4]
```diff
@@ -897,6 +897,94 @@ func TestValidateKeyToPath(t *testing.T) {
...
+			errtype:  field.ErrorTypeRequired,
+			errfield: "endpoints",
+		},
+		{
+			name:     "missing path",
+			nfs:      &api.GlusterfsVolumeSource{EndpointsName: "my-ebdpoint", Path: ""},
```
Comment (rootfs) [inline pkg/api/validation/validation_test.go]: my-endpoint

File: pkg/api/validation/validation_test.go [hunk 5]
```diff
@@ -897,6 +897,94 @@ func TestValidateKeyToPath(t *testing.T) {
...
+			errfield: "path",
+		},
+	}
+
+	for i, tc := range testCases {
+		errs := validateGlusterfs(tc.nfs, field.NewPath("field"))
+
+		if len(errs) > 0 && tc.errtype == "" {
+			t.Errorf("[%d: %q] unexpected error(s): %v", i, tc.name, errs)
+		} else if len(errs) > 1 {
```
Comment (rootfs) [inline pkg/api/validation/validation_test.go]: why are 2-error test cases impossible (if you add one)?

File: pkg/api/validation/validation_test.go [hunk 6]
```diff
@@ -897,6 +897,94 @@ func TestValidateKeyToPath(t *testing.T) {
...
+			errtype:  field.ErrorTypeRequired,
+			errfield: "endpoints",
+		},
+		{
+			name:     "missing path",
+			nfs:      &api.GlusterfsVolumeSource{EndpointsName: "my-ebdpoint", Path: ""},
+			errtype:  field.ErrorTypeRequired,
+			errfield: "path",
+		},
+	}
```
Comment (rootfs) [inline pkg/api/validation/validation_test.go]: can there be a test with both endpoint and path missing?

**Comment by rootfs** [comment]
@zjj2wry thanks!

**Comment by zjj2wry** [comment]
@rootfs think for your careful review. 😊 i make mary mistakes. please check it agains. think you ~

**Pull request merged**